### ISSUE TITLE
Retriggering jobs with triggerbyfilters.py is slow and does a lot of unnecessary requests

### ISSUE THREAD

**Issue by jgraham** [armenzg/mozilla_ci_tools#200]
When running this script I seem to be doing dozens of http requests and repeatedly downloading builds-4hr.json (amongst other things, I suspect).

So in an ideal world I imagine that this would grab all the data about the jobs that have already been scheduled for the revision from treeherder and then post a single message to treeherder specifying the new jobs it ought to schedule. 

I appreciate that we are far from that world.

In the interim, it would be great if mozci had the ability to populate some sort of State object that could be passed around. Then scripts like this which trigger multiple things in a loop could populate the state once, outside the loop, and then reuse that state for all subsequent operations. Of course the state may become outdated in that time, but that doesn't seem like a big problem for this use case.

**Comment by vaibhavmagarwal** [comment]
One important corner case to remember here is that suppose for a particular test the build is not present, then if we trigger the build, we need to download the latest builds_4hr.js again else reusing the old builds_4hr.js would result in triggering the build multiple times.

**Comment by armenzg** [comment]
Unless we could look into builds-running.js and builds-pending.js to ligthen this OR TH status.

**Comment by adusca** [comment]
What about downloading builds_4hrs once, deciding what needs to be triggered using that and then trigger everything without downloading again, just assuming the builds we triggered were actually triggered?

**Comment by armenzg** [comment]
If we fix this, we will also make alltalos.py infinitely faster (I think).

For every buildername, we are determining a trigger objective [1].
Every time we call query_jobs(repo_name, revision) we grab buildername after buildername very similar job status information. In many cases new info about jobs that we have triggered on a previous run of the loop.

I think that we can keep the state of the first query_jobs() fetch and only fetch newer data if we believe that we need it to trigger a build job [2] instead of a test job (I would rather grab fresh data to avoid having to trigger a build job that might have *just* been scheduled). Nevertheless, this case would happen only once. I might be overthinking since (read below) I added measures to prevent this.

We have a way to prevent scheduling build jobs unnecessarily [3]. This keeps state in some sense.

We could make query_jobs() follow a similar pattern and avoid hitting builds-4hr.js unnecessarily.
We could add an optional parameter to query_jobs() to force a fetch of status or a method to reset the state (if we need to).

In other words, we all seem to be on the same page.

[1] https://github.com/armenzg/mozilla_ci_tools/blob/master/mozci/mozci.py#L85
[2] https://github.com/armenzg/mozilla_ci_tools/blob/master/mozci/mozci.py#L167
[3] https://github.com/armenzg/mozilla_ci_tools/blob/master/mozci/mozci.py#L39

**Comment by adusca** [comment]
I believe a good way to start dealing with this is making _determine_trigger_objective deal with a list of buildernames for the same revision instead of just one. It could then return one tuple of builder_to_trigger, file per buildername in the list.

About lowering the number of post requests we do, I'm looking at buildapi self-serve documentation [1]. It looks like it's mostly one post request per job. I'm curious about this part:

POST	/self-serve/{branch}/build	Rebuild `build_id`, which must be passed in as a POST parameter.`priority` and `count` are also accepted as optional parameters.`count` defaults to 1, and represents the number of times this build will be rebuilt.

Does that means that we can somehow pass 'count' to buildpapi self-serve to retrigger a job N times instead of doing N post requests?

[1] https://secure.pub.build.mozilla.org/buildapi/self-serve

**Comment by jgraham** [comment]
> Does that means that we can somehow pass 'count' to buildpapi self-serve to retrigger a job N times instead of doing N post requests?

Yes. See the buildapi UI e..g [1]

[1] https://secure.pub.build.mozilla.org/buildapi/self-serve/mozilla-inbound/rev/aa8ad4978b15

**Comment by armenzg** [comment]
I believe you got it right. I have not tried it myself.

What you mention about _determine_trigger_objective could also work, however, I think being to self-manage previously fetched data about job status might be better since we would not have to re-design all other functions that call _determine_trigger_objective().
I have not thought deeply into this so feel free to not discard your proposal.

**Comment by armenzg** [comment]
I believe @adusca improved this dramatically with https://github.com/armenzg/mozilla_ci_tools/pull/221
and https://github.com/armenzg/mozilla_ci_tools/pull/220.
Blog post: http://explique.me/cProfile
